src/main/java/com/snowflake/core/util/HiveSyncTool.java

```diff
@@ -0,0 +1,232 @@
+  //       wrong results, e.g. client.getAllDatabases() returns no databases.
+  private final HiveMetaStoreClient hmsClient;
+
+  // The Snowflake configuration
+  private final SnowflakeConf snowflakeConf;
+
+  /**
+   * Instantiates the HiveSyncTool.
+   * @param hmsClient a client for the Hive metastore
+   */
+  public HiveSyncTool(HiveMetaStoreClient hmsClient)
+  {
+    this.hmsClient = hmsClient;
+    this.snowflakeConf = new SnowflakeConf();
+  }
+
+  /**
+   * Does a one-time, one-way metadata sync from the Hive metastore to
+   * Snowflake.
+   * 1. Tables in Hive are created in Snowflake
+   * 2. Partitions not in Hive are dropped from Snowflake
+   * 3. Partitions in Hive are added to Snowflake
+   * Note: does not drop tables from Snowflake.
+   * @throws TException thrown when encountering a Thrift exception while
+   *         communicating with the metastore or executing a metastore operation
+   */
+  public void sync() throws TException
+  {
+    Pattern tableNameFilter = snowflakeConf.getPattern(
+        SnowflakeConf.ConfVars.SNOWFLAKE_TABLE_FILTER_REGEX.getVarname(), null);
+    Pattern databaseNameFilter = snowflakeConf.getPattern(
+        SnowflakeConf.ConfVars.SNOWFLAKE_DATABASE_FILTER_REGEX.getVarname(), null);
+
+    log.info("Starting sync");
+    List<String> databaseNames = hmsClient.getAllDatabases().stream()
+        .filter(db -> databaseNameFilter == null || !databaseNameFilter.matcher(db).matches())
+        .collect(Collectors.toList());
+    log.info(String.format("Syncing %s databases from Hive",
+                           databaseNames.size()));
+    for (String databaseName : databaseNames)
+    {
+      Preconditions.checkNotNull(databaseName);
+      List<String> tableNames = hmsClient.getAllTables(databaseName).stream()
+          .filter(table -> tableNameFilter == null || !tableNameFilter.matcher(table).matches())
+          .collect(Collectors.toList());
+      log.info(String.format("Syncing %s tables for database %s",
+                             tableNames.size(), databaseName));
+      for (String tableName : tableNames)
+      {
+        Preconditions.checkNotNull(tableName);
+
+        // Do a touch on the table to fire listener events
+        Table hiveTable = hmsClient.getTable(databaseName, tableName);
+        hmsClient.alter_table(databaseName, tableName, hiveTable);
+
+        if (!hiveTable.getPartitionKeys().isEmpty())
+        {
+          // Drop extra partitions
+          dropExtraPartitionsFromSnowflake(databaseName, hiveTable);
+        }
+
+        // Do a touch on all partitions to fire listener events
+        List<Partition> partitions = hmsClient.listPartitions(
+            databaseName, tableName, (short) -1 /* all partitions */);
+        log.info(String.format("Syncing %s partitions for table %s.%s",
+                               partitions.size(), tableName, databaseName));
+        hmsClient.alter_partitions(databaseName, tableName, partitions);
```

Comment:
is this the documented way to touch objects in hive? Is there a way to generate specific events?

Reply:
I was going to change this to fire our own commands directly, but forgot to make that change. It'd be pretty dangerous otherwise.

Reply:
It's updated now. We don't generate events per se, but we'll generate the commands which is basically the same thing.